file/file_prefetch_buffer.h

```diff
@@ -487,10 +488,104 @@ class FilePrefetchBuffer {
+                            uint64_t end_offset1, size_t alignment,
+                            size_t readahead_size);
+
+  // *** BEGIN Helper APIs related to data in Buffers ***
+  bool IsDataBlockInBuffer(BufferInfo* buf, uint64_t offset, size_t length) {
+    return (offset >= buf->offset_ &&
+            offset + length <= buf->offset_ + buf->buffer_.CurrentSize());
+  }
+  bool IsOffsetInBuffer(BufferInfo* buf, uint64_t offset) {
+    return (offset >= buf->offset_ &&
+            offset < buf->offset_ + buf->buffer_.CurrentSize());
+  }
+  bool DoesBufferContainData(BufferInfo* buf) {
+    return buf->buffer_.CurrentSize() > 0;
+  }
+  bool IsBufferOutdated(BufferInfo* buf, uint64_t offset) {
+    return (!buf->async_read_in_progress_ && DoesBufferContainData(buf) &&
+            offset >= buf->offset_ + buf->buffer_.CurrentSize());
+  }
+  bool IsBufferOutdatedWithAsyncProgress(BufferInfo* buf, uint64_t offset) {
+    return (buf->async_read_in_progress_ && buf->io_handle_ != nullptr &&
+            offset >= buf->offset_ + buf->async_req_len_);
+  }
+  bool IsOffsetInBufferWithAsyncProgress(BufferInfo* buf, uint64_t offset) {
+    return (buf->async_read_in_progress_ && offset >= buf->offset_ &&
+            offset < buf->offset_ + buf->async_req_len_);
+  }
+  // *** END Helper APIs related to data in Buffers ***
+
+  // *** BEGIN APIs related to allocating and freeing buffers ***
+  bool IsBufferQueueEmpty() { return bufs_.empty(); }
+
+  BufferInfo* GetFirstBuffer() { return bufs_.front(); }
+
+  BufferInfo* GetLastBuffer() { return bufs_.back(); }
+
+  size_t NumBuffersAllocated() { return bufs_.size(); }
+
+  void AllocateBuffer() {
+    assert(!free_bufs_.empty());
+    BufferInfo* buf = free_bufs_.front();
+    free_bufs_.pop_front();
+    bufs_.emplace_back(buf);
+  }
+
+  void AllocateBufferIfEmpty() {
+    if (bufs_.empty()) {
+      AllocateBuffer();
+    }
+  }
+
+  void FreeFrontBuffer() {
+    BufferInfo* buf = bufs_.front();
+    buf->buffer_.Clear();
+    bufs_.pop_front();
+    free_bufs_.emplace_back(buf);
+  }
+
+  void FreeLastBuffer() {
+    BufferInfo* buf = bufs_.back();
+    buf->buffer_.Clear();
+    bufs_.pop_back();
+    free_bufs_.emplace_back(buf);
+  }
+
+  void FreeAllBuffers() {
+    for (auto& buf : bufs_) {
+      buf->ClearBuffer();
+      bufs_.pop_front();
+      free_bufs_.emplace_back(buf);
+    }
+  }
+
+  void FreeEmptyBuffers() {
+    if (bufs_.empty()) {
+      return;
+    }
+
+    std::deque<BufferInfo*> tmp_buf;
+    while (!bufs_.empty()) {
+      BufferInfo* buf = bufs_.front();
+      bufs_.pop_front();
+      if (buf->async_read_in_progress_ || DoesBufferContainData(buf)) {
+        tmp_buf.emplace_back(buf);
+      } else {
+        free_bufs_.emplace_back(buf);
+      }
+    }
+    bufs_ = tmp_buf;
+  }
+
+  // *** END APIs related to allocating and freeing buffers ***
+
+  std::deque<BufferInfo*> bufs_;
+  std::deque<BufferInfo*> free_bufs_;
+  std::deque<BufferInfo*> overlap_bufs_;
```

Comment:
Deque works better because it's easy to remove from the front and back. Let me see If I can make it queue since now implementation has been finalized. Let me know If you have any concerns with Queue.

Reply:
Oh sorry I meant to comment only on `overlap_bufs_`. GitHub shows the context so it might have looked like I was commenting on all three uses of `deque`. The reason I thought `overlap_bufs_` may not need to be a `deque` is because it appears to always have zero or one entry, in which case it wouldn't need to be a collection data type.

Reply:
Ohh okk! Yeah that makes sense. I can change that. Thanks